apsuite/commisslib/measure_bbb_data.py

```diff
@@ -523,3 +523,250 @@ def waterfall_plot(axis, xs, zs, data):
+            'timestamp': _time.time(),
+            'stored_current': bbb.dcct.current,
+            'data': sb_tune.data_raw,
+            'spec_mag': sb_tune.spec_mag, 'spec_freq': sb_tune.spec_freq,
+            'spec_phase': sb_tune.spec_phase,
+            'bunch_id': sb_tune.bunch_id, 'fft_size': sb_tune.fft_size,
+            'fft_overlap': sb_tune.fft_overlap,
+            'delay_calibration': sb_tune.delay_calibration,
+            'nr_averages': sb_tune.nr_averages}
+        return data
+
+    def merge_data(data1, data2):
+        """."""
+        if data1.keys() != data2.keys():
+            raise Exception('Incompatible datas')
+        merge = dict()
+        for key in data1:
+            merge[key] = data1[key] + data2[key]
+        return merge
+
+    def turn_on_pingers_pulse(self):
+        """."""
+        for ping in self.pingers:
+            ping.cmd_turn_on_pulse()
+
+    def turn_off_pingers_pulse(self):
+        """."""
+        for ping in self.pingers:
+            ping.cmd_turn_off_pulse()
+
+    def turn_on_pingers(self):
+        """."""
+        for ping in self.pingers:
+            ping.cmd_turn_on()
+
+    def turn_off_pingers(self):
+        """."""
+        for ping in self.pingers:
+            ping.cmd_turn_off()
+
+    def prepare_pingers(self):
+        """."""
+        self.devices['pingh'].strength = self.params.kickh
+        self.devices['pingv'].strength = self.params.kickv
+        self.turn_on_pingers()
+        self.turn_on_pingers_pulse()
+
+    def inject_in_storage_ring(self, goal_curr):
+        """."""
+        self.devices['egun'].cmd_enable_trigger()
+        while not self._check_stored_current(goal_curr):
+            _time.sleep(TuneShiftParams.WAIT_INJ)
+        curr = self.devices['bbbh'].dcct.current
+        print(
+            f'Stored Current: {curr:.3f}/{goal_curr:.3f}mA.')
+        self.devices['egun'].cmd_disable_trigger()
+
+    def _check_stored_current(
+            self, goal_curr, tol=TuneShiftParams.DEF_TOL_CURRENT):
+        dcct_curr = self.devices['bbbh'].dcct.current
+        return dcct_curr > goal_curr or abs(dcct_curr - goal_curr) < tol
+
+    def _check_pingers_problem(self):
+        for ping in self.pingers:
+            if ping.voltage_mon < 0:
+                # reset pinger
+                ping.cmd_turn_off()
+                ping.cmd_turn_on()
+                return True
+        return False
+
+    def run_meas(self, save=True):
+        """."""
+        data = dict()
+        datah = list()
+        datav = list()
+        currs = list()
+        for curr in self.params.currents:
+            t0 = _time.time()
+            self.inject_in_storage_ring(curr)
+
+            trial = 0
+            while self._check_pingers_problem():
+                if trial > 2:
+                    print('3 unsucessful reset trials. Exiting...')
+                    break
+                print('Problem with pingers voltage. Resetting...')
+                trial += 1
```

Comment:
Maybe it is prudent to add a `sleep` here. What do you think?

Reply:
The turn on/off commands for pingers already have `wait` methods, but I agree it is prudent to wait a little bit (for example when the pingers voltages are still increasing after turning them on). Done!